less_users/views.py

```diff
@@ -75,69 +84,112 @@ def get_queryset(self):
-        user_new_challenge.save()
-        messages.success(request, 'You activated new challenge!')
-        return HttpResponseRedirect(reverse('challenge_detail', args=[str(pk)]))
+    user_challenge = UserChallenge.objects.filter(
+        user=request.user, challenge_id=pk, is_active=True
+    )
+    if user_challenge.exists():
+        messages.warning(request, "You still have this challenge active!")
+        return redirect("my_challenges")
     else:
-        messages.warning(request, 'You still have this challenge active!')
-        return redirect('my_challenges')
+        user_new_challenge = UserChallenge.objects.create(
+            user=request.user, challenge_id=pk
+        )
+        messages.success(request, "You activated new challenge!")
+        return HttpResponseRedirect(reverse("challenge_detail", args=[str(pk)]))
 
 
-def event_view(request, **kwargs):
-    """gain points when challenge is completed; stop challenge or detele challenge """
+def event_view_done(request):
+    """gain points when user challenge is completed by clicking >done< button;
+    should check if points can be added (depending on duration and frequency in model)"""
     user_id = request.user
-    if 'done' in request.POST:
-        """should check if points can be added (depending on durration and frequncy in model"""
-        challenge_id = int(request.POST.get('done'))
-        points = Challenge.objects.get(id=challenge_id).points
-        this = UserChallenge.objects.get(user_id=user_id, challenge_id=challenge_id, is_active=True)
-        user_challenge_id = this.id
-        this.check_if_active()
-        if this.get_points == points and this.is_active:
-            new_log = Log.objects.create(user_challenge_id=user_challenge_id, points=points)
-            my_profile = Profile.objects.get(user_id=user_id)
-            my_profile.points += points
-            my_profile.save()
-            messages.success(request, f'You have got new points {points} for challenge: {this.challenge.name}!')
-        elif this.get_points == 0 or this.is_active == False:
-            messages.warning(request, f'You cannot get new points for this challenge yet: {this.challenge.name}!')
-    elif 'stop' in request.POST:
-        user_challenge_id = int(request.POST.get('stop'))
-        this = UserChallenge.objects.get(id=user_challenge_id)
-        this.is_active = False
-        this.save()
-        messages.info(request, f'You stopped challenge: {this.challenge.name}. You can activate the new one.')
-    elif 'delete' in request.POST:
-        user_challenge_id = int(request.POST.get('delete'))
-        this = UserChallenge.objects.get(id=user_challenge_id)
-        this.delete()
-        messages.warning(request, f'You have deleted {this.challenge.name} from your challenges')
-    return redirect('my_challenges')
-
-
-
+    challenge_id = int(request.POST.get("done"))
+    points = Challenge.objects.get(id=challenge_id).points
```

Comment:
Co jeżeli challenge nie istnieje?

Reply:
przyciski "done" odnoszą się do każdego wyzwania, a ich id są pobierane z formularza post. Myślisz, że da się z ręki wpisać adres strony żeby aktywować challenge, żeby mogło go nie być? hm... jeśli tak, to musiałabym sprawdzić jeszcze warunek, że istnieje.

Reply:
Aha, może być jeszcze ...objects.get_or_404 - o to Ci chodzi?
Albo znów sprawdzić objects.exists(), a potem dalej  z kodem.

Reply:
> Myślisz, że da się z ręki wpisać adres strony żeby aktywować challenge, żeby mogło go nie być? 

Można zmodyfikować HTML na stronie, można zrobić request w przeglądarce, terminalu, postmanie czy innym programie, który pozwala na wysyłanie requestów.

Chodzi mi o to, że w kodzie nie można zawsze zakładać, że wszystko pójdzie dobrze a użytkownik nigdy się nie pomyli. Zaznaczyłam tutaj tylko kilka wybiórczych przykładów kiedy coś może pójść źle, ale generalnie to pisząc każdą linjkę musisz się zastanowić czy coś może pójść nie tak? Co jeżeli użytkownik prześle złe dane? Co jeżeli ktoś wyśle złe zapytanie? Co jeżeli obiekt został usunięty pomiędzy wyrenderowaniem strony a wysłaniem requesta? A jak już wymyślisz takie przykłady to obsłużyć je w kodzie i napisać testy, żeby się upewnić.